fdbclient/ClientStatusReport.cpp

```diff
@@ -0,0 +1,173 @@
+/*
+ * ClientStatusReport.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbclient/CommitProxyInterface.h"
+#include "fdbclient/CoordinationInterface.h"
+#include "fdbclient/DatabaseContext.h"
+#include "fdbrpc/FlowTransport.h"
+#include "fdbclient/json_spirit/json_spirit_value.h"
+#include "fdbclient/json_spirit/json_spirit_writer_template.h"
+#include <set>
+
+namespace {
+
+class ClientReportGenerator {
+public:
+	ClientReportGenerator(DatabaseContext& cx) : cx(cx), healthy(true), numConnectionsFailed(0) {}
+
+	Standalone<StringRef> generateReport() {
+		if (cx.isError()) {
+			statusObj["InitializationError"] = cx.deferredError.code();
+			healthy = false;
+		} else {
+			reportCoordinators();
+			reportClientInfo();
+			reportStorageServers();
+			reportConnections();
+			statusObj["Healthy"] = healthy;
+		}
+		return StringRef(json_spirit::write_string(json_spirit::mValue(statusObj)));
+	}
+
+private:
+	void reportCoordinators() {
+		auto& connRecord = cx.connectionRecord->get();
+		ClusterConnectionString cs = connRecord->getConnectionString();
+		json_spirit::mArray coordArray;
+		for (const auto& hostName : cs.hostnames) {
+			coordArray.push_back(hostName.toString());
+		}
+		for (const auto& addr : cs.coords) {
+			coordArray.push_back(addr.toString());
+			serverAddresses.insert(addr);
+		}
+		statusObj["Coordinators"] = coordArray;
+		if (cx.coordinator->get().present()) {
+			statusObj["CurrentCoordinator"] = cx.coordinator->get().get().getAddressString();
+		}
+
+		// Update health status
+		if (cs.hostnames.size() + cs.coords.size() == 0) {
+			healthy = false;
+		}
+		if (!cx.coordinator->get().present()) {
+			healthy = false;
+		}
+	}
+
+	void reportClientInfo() {
+		auto& clientInfo = cx.clientInfo->get();
+		statusObj["ClusterID"] = clientInfo.clusterId.toString();
```

Comment:
There is a ClusterId reported from the Muti-Version Database. I think I will remove that one, leaving only the Cluster ID from the DatabaseContext, because in MVC is not always initialized.